File: apps/ios/GalaxySSI/AgentCronExpression.swift
import Foundation

enum AgentCronExpressionError: LocalizedError, Equatable {
  case invalid(String)

  var errorDescription: String? {
    switch self {
    case .invalid(let detail):
      return detail
    }
  }
}

struct AgentCronExpression: Equatable {
  let expression: String
  private let minute: Field
  private let hour: Field
  private let day: Field
  private let month: Field
  private let weekday: Field

  func matches(date: Date, timeZone: TimeZone) -> Bool {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    let components = calendar.dateComponents([.minute, .hour, .day, .month, .weekday], from: date)
    let cronWeekday = ((components.weekday ?? 1) - 1) % 7
    let dayMatches = day.values.contains(components.day ?? -1)
    let weekdayMatches = weekday.values.contains(cronWeekday)
    let calendarMatches: Bool
    if day.wildcard && weekday.wildcard {
      calendarMatches = true
    } else if day.wildcard {
      calendarMatches = weekdayMatches
    } else if weekday.wildcard {
      calendarMatches = dayMatches
    } else {
      calendarMatches = dayMatches || weekdayMatches
    }
    return minute.values.contains(components.minute ?? -1) &&
      hour.values.contains(components.hour ?? -1) &&
      month.values.contains(components.month ?? -1) &&
      calendarMatches
  }

  func nextAfter(timestampMillis: Int64, timeZoneIdentifier: String) throws -> Int64 {
    let timeZone = try Self.parseZone(timeZoneIdentifier)
    var candidate = floorToMinute(timestampMillis: timestampMillis, timeZone: timeZone, minuteOffset: 1)
    for _ in 0..<Self.maxScanMinutes {
      if matches(date: candidate, timeZone: timeZone) {
        return Self.millis(candidate)
      }
      candidate = candidate.addingTimeInterval(60)
    }
    throw AgentCronExpressionError.invalid("Cron has no occurrence within six years")
  }

  func previousAtOrBefore(timestampMillis: Int64, timeZoneIdentifier: String) throws -> Int64 {
    let timeZone = try Self.parseZone(timeZoneIdentifier)
    var candidate = floorToMinute(timestampMillis: timestampMillis, timeZone: timeZone, minuteOffset: 0)
    for _ in 0..<Self.maxScanMinutes {
      if matches(date: candidate, timeZone: timeZone) {
        return Self.millis(candidate)
      }
      candidate = candidate.addingTimeInterval(-60)
    }
    throw AgentCronExpressionError.invalid("Cron has no occurrence within six years")
  }

  static func parse(_ expression: String) throws -> AgentCronExpression {
    let parts = expression
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .components(separatedBy: .whitespacesAndNewlines)
      .filter { !$0.isEmpty }
    guard parts.count == 5 else {
      throw AgentCronExpressionError.invalid("Cron requires five fields")
    }
    return AgentCronExpression(
      expression: parts.joined(separator: " "),
      minute: try parseField(parts[0], minimum: 0, maximum: 59),
      hour: try parseField(parts[1], minimum: 0, maximum: 23),
      day: try parseField(parts[2], minimum: 1, maximum: 31),
      month: try parseField(parts[3], minimum: 1, maximum: 12, aliases: monthNames),
      weekday: try parseField(parts[4], minimum: 0, maximum: 6, aliases: weekdayNames, sundayAlias: true)
    )
  }

  static func parseZone(_ timeZoneIdentifier: String) throws -> TimeZone {
    let clean = timeZoneIdentifier.trimmingCharacters(in: .whitespacesAndNewlines)
    let identifier = clean.isEmpty ? "UTC" : clean
    guard let zone = TimeZone(identifier: identifier) else {
      throw AgentCronExpressionError.invalid("Unknown time zone: \(timeZoneIdentifier)")
    }
    return zone
  }

  private func floorToMinute(timestampMillis: Int64, timeZone: TimeZone, minuteOffset: Int) -> Date {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    let date = Date(timeIntervalSince1970: Double(timestampMillis) / 1_000)
    var components = calendar.dateComponents([.year, .month, .day, .hour, .minute], from: date)
    components.second = 0
    components.nanosecond = 0
    let floored = calendar.date(from: components) ?? date
    return calendar.date(byAdding: .minute, value: minuteOffset, to: floored) ?? floored.addingTimeInterval(Double(minuteOffset) * 60)
  }

  private static func parseField(
    _ text: String,
    minimum: Int,
    maximum: Int,
    aliases: [String: Int] = [:],
    sundayAlias: Bool = false
  ) throws -> Field {
    let clean = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !clean.isEmpty else {
      throw AgentCronExpressionError.invalid("Cron field is blank")
    }
    var output = Set<Int>()
    for clause in clean.split(separator: ",", omittingEmptySubsequences: false).map(String.init) {
      let parts = clause.split(separator: "/", maxSplits: 1, omittingEmptySubsequences: false).map(String.init)
      let base = parts[0]
      let step: Int
      if parts.count == 2 {
        guard let value = Int(parts[1]) else {
          throw AgentCronExpressionError.invalid("Invalid cron step")
        }
        step = value
      } else {
        step = 1
      }
      guard step > 0 else {
        throw AgentCronExpressionError.invalid("Cron step must be positive")
      }
      let range: ClosedRange<Int>
      if base == "*" {
        range = minimum...maximum
      } else if base.contains("-") {
        let edges = base.split(separator: "-", maxSplits: 1, omittingEmptySubsequences: false).map(String.init)
        guard edges.count == 2 else {
          throw AgentCronExpressionError.invalid("Invalid cron range")
        }
        let lower = try numeric(edges[0], aliases: aliases, sundayAlias: sundayAlias)
        let upper = try numeric(edges[1], aliases: aliases, sundayAlias: sundayAlias)
        guard upper >= lower else {
          throw AgentCronExpressionError.invalid("Cron ranges cannot wrap")
        }
        range = lower...upper
      } else {
        let value = try numeric(base, aliases: aliases, sundayAlias: sundayAlias)
        range = value...value
      }
      guard range.lowerBound >= minimum && range.upperBound <= maximum else {
        throw AgentCronExpressionError.invalid("Cron value is outside \(minimum)..\(maximum)")
      }
      for value in stride(from: range.lowerBound, through: range.upperBound, by: step) {
        output.insert(value)
      }
    }
    guard !output.isEmpty else {
      throw AgentCronExpressionError.invalid("Cron field has no values")
    }
    return Field(values: output, wildcard: clean == "*")
  }

  private static func numeric(
    _ raw: String,
    aliases: [String: Int],
    sundayAlias: Bool
  ) throws -> Int {
    let clean = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    if let alias = aliases[clean] {
      return alias
    }
    guard let value = Int(clean) else {
      throw AgentCronExpressionError.invalid("Invalid cron token: \(raw)")
    }
    return sundayAlias && value == 7 ? 0 : value
  }

  private static func millis(_ date: Date) -> Int64 {
    Int64((date.timeIntervalSince1970 * 1_000).rounded())
  }

  private struct Field: Equatable {
    var values: Set<Int>
    var wildcard: Bool
  }

  private static let maxScanMinutes = 3_200_000
  private static let monthNames = [
    "jan": 1, "feb": 2, "mar": 3, "apr": 4, "may": 5, "jun": 6,
    "jul": 7, "aug": 8, "sep": 9, "oct": 10, "nov": 11, "dec": 12
  ]
  private static let weekdayNames = [
    "sun": 0, "mon": 1, "tue": 2, "wed": 3, "thu": 4, "fri": 5, "sat": 6
  ]
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentCronExpression.kt
package com.galaxyssi.chat

import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

class AgentCronExpression private constructor(
    val expression: String,
    private val minute: Field,
    private val hour: Field,
    private val day: Field,
    private val month: Field,
    private val weekday: Field
) {
    fun matches(value: ZonedDateTime): Boolean {
        val cronWeekday = value.dayOfWeek.value % 7
        val dayMatches = value.dayOfMonth in day.values
        val weekdayMatches = cronWeekday in weekday.values
        val calendarMatches = when {
            day.wildcard && weekday.wildcard -> true
            day.wildcard -> weekdayMatches
            weekday.wildcard -> dayMatches
            else -> dayMatches || weekdayMatches
        }
        return value.minute in minute.values &&
            value.hour in hour.values &&
            value.monthValue in month.values &&
            calendarMatches
    }

    fun nextAfter(timestampMillis: Long, zoneId: String): Long {
        val zone = parseZone(zoneId)
        var candidate = Instant.ofEpochMilli(timestampMillis)
            .atZone(zone)
            .withSecond(0)
            .withNano(0)
            .plusMinutes(1)
        repeat(MAX_SCAN_MINUTES) {
            if (matches(candidate)) return candidate.toInstant().toEpochMilli()
            candidate = candidate.plusMinutes(1)
        }
        throw IllegalArgumentException("Cron has no occurrence within six years")
    }

    fun previousAtOrBefore(timestampMillis: Long, zoneId: String): Long {
        val zone = parseZone(zoneId)
        var candidate = Instant.ofEpochMilli(timestampMillis)
            .atZone(zone)
            .withSecond(0)
            .withNano(0)
        repeat(MAX_SCAN_MINUTES) {
            if (matches(candidate)) return candidate.toInstant().toEpochMilli()
            candidate = candidate.minusMinutes(1)
        }
        throw IllegalArgumentException("Cron has no occurrence within six years")
    }

    private data class Field(
        val values: Set<Int>,
        val wildcard: Boolean
    )

    companion object {
        private const val MAX_SCAN_MINUTES = 3_200_000
        private val MONTH_NAMES = mapOf(
            "jan" to 1,
            "feb" to 2,
            "mar" to 3,
            "apr" to 4,
            "may" to 5,
            "jun" to 6,
            "jul" to 7,
            "aug" to 8,
            "sep" to 9,
            "oct" to 10,
            "nov" to 11,
            "dec" to 12
        )
        private val WEEKDAY_NAMES = mapOf(
            "sun" to 0,
            "mon" to 1,
            "tue" to 2,
            "wed" to 3,
            "thu" to 4,
            "fri" to 5,
            "sat" to 6
        )

        fun parse(expression: String): AgentCronExpression {
            val parts = expression.trim().split(Regex("\\s+"))
            require(parts.size == 5) { "Cron requires five fields" }
            return AgentCronExpression(
                expression = parts.joinToString(" "),
                minute = parseField(parts[0], 0, 59),
                hour = parseField(parts[1], 0, 23),
                day = parseField(parts[2], 1, 31),
                month = parseField(parts[3], 1, 12, MONTH_NAMES),
                weekday = parseField(parts[4], 0, 6, WEEKDAY_NAMES, sundayAlias = true)
            )
        }

        fun parseZone(zoneId: String): ZoneId = runCatching {
            ZoneId.of(zoneId.ifBlank { "UTC" })
        }.getOrElse {
            throw IllegalArgumentException("Unknown time zone: $zoneId")
        }

        private fun parseField(
            text: String,
            minimum: Int,
            maximum: Int,
            aliases: Map<String, Int> = emptyMap(),
            sundayAlias: Boolean = false
        ): Field {
            val clean = text.trim().lowercase()
            require(clean.isNotBlank()) { "Cron field is blank" }
            val output = linkedSetOf<Int>()
            clean.split(",").forEach { clause ->
                val parts = clause.split("/", limit = 2)
                val base = parts[0]
                val step = if (parts.size == 2) {
                    parts[1].toIntOrNull() ?: throw IllegalArgumentException("Invalid cron step")
                } else {
                    1
                }
                require(step > 0) { "Cron step must be positive" }
                val range = when {
                    base == "*" -> minimum..maximum
                    "-" in base -> {
                        val edges = base.split("-", limit = 2)
                        numeric(edges[0], aliases, sundayAlias)..numeric(edges[1], aliases, sundayAlias)
                    }
                    else -> {
                        val value = numeric(base, aliases, sundayAlias)
                        value..value
                    }
                }
                require(range.first in minimum..maximum && range.last in minimum..maximum) {
                    "Cron value is outside $minimum..$maximum"
                }
                require(range.last >= range.first) { "Cron ranges cannot wrap" }
                range.step(step).forEach(output::add)
            }
            require(output.isNotEmpty()) { "Cron field has no values" }
            return Field(output, clean == "*")
        }

        private fun numeric(
            raw: String,
            aliases: Map<String, Int>,
            sundayAlias: Boolean
        ): Int {
            val clean = raw.trim().lowercase()
            aliases[clean]?.let { return it }
            val value = clean.toIntOrNull() ?: throw IllegalArgumentException("Invalid cron token: $raw")
            return if (sundayAlias && value == 7) 0 else value
        }
    }
}
